pkg/util/mount/nsenter_mount.go

```diff
@@ -281,49 +273,179 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+	// Bind-mount the subpath to avoid using symlinks in subpaths.
+	newHostPath, err = doNsEnterBindSubPath(mounter, subPath)
+
+	// There is no action when the container starts. Bind-mount will be cleaned
+	// when container stops by CleanSubPaths.
+	cleanupAction = nil
+	return newHostPath, cleanupAction, err
+}
+
+func (mounter *NsenterMounter) SafeMakeDir(subdir string, base string, perm os.FileMode) error {
+	fullSubdirPath := filepath.Join(base, subdir)
+	evaluatedSubdirPath, err := mounter.evalHostSymlinks(fullSubdirPath, false /* mustExist */)
 	if err != nil {
-		return 0, fmt.Errorf("error reading %s: %s", procStatusPath, err)
+		return fmt.Errorf("error resolving symlinks in %s: %s", fullSubdirPath, err)
 	}
-	matches := pidRegExp.FindSubmatch(statusBytes)
-	if len(matches) < 2 {
-		return 0, fmt.Errorf("cannot parse %s: no Pid:", procStatusPath)
+	kubeletSubdirPath := mounter.getKubeletPath(evaluatedSubdirPath)
+
+	evaluatedBase, err := mounter.evalHostSymlinks(base, true /* mustExist */)
+	if err != nil {
+		return fmt.Errorf("error resolving symlinks in %s: %s", base, err)
 	}
-	return strconv.Atoi(string(matches[1]))
+	kubeletBase := mounter.getKubeletPath(evaluatedBase)
+
+	return doSafeMakeDir(kubeletSubdirPath, kubeletBase, perm)
 }
 
-func (mounter *NsenterMounter) PrepareSafeSubpath(subPath Subpath) (newHostPath string, cleanupAction func(), err error) {
-	hostPid, err := mounter.getPidOnHost(hostProcSelfStatusPath)
+func (mounter *NsenterMounter) getKubeletPath(path string) string {
+	return filepath.Join(nsenter.HostRootFsPath, path)
+}
+
+func (mounter *NsenterMounter) Lstat(path string) (os.FileInfo, error) {
+	// Resolve any symlinks on the host
+	evaluatedPath, err := mounter.evalHostSymlinks(path, true /* mustExist */)
 	if err != nil {
-		return "", nil, err
+		return nil, err
 	}
-	glog.V(4).Infof("Kubelet's PID on the host is %d", hostPid)
+	// But call Lstat in kubelet
+	evaluatedKubeletPath := mounter.getKubeletPath(evaluatedPath)
+	return os.Lstat(evaluatedKubeletPath)
+}
 
-	// Bind-mount the subpath to avoid using symlinks in subpaths.
-	newHostPath, err = doBindSubPath(mounter, subPath, hostPid)
+// evalHostSymlinks returns the path name after the evaluation of any symbolic
+// links on the host. In other words, it does the same as filepath.EvalSymlinks()
+// would do when it ran on the host and not inside a container.
+//
+// mustExist makes evalHostSymlinks to return error when the path does not
+// exist. When it's false, it evaluates symlinks of the existing part and
+// blindly adds the non-existing part:
+// path: /mnt/volume/non/existing/directory. /mnt/volume exists and
+// non/existing/directory does not exist. It resolves symlinks in /mnt/volume
+// to say /mnt/foo and returns /mnt/foo/non/existing/directory.
+func (mounter *NsenterMounter) evalHostSymlinks(path string, mustExist bool) (string, error) {
```

Comment:
I had added a similar function in https://github.com/kubernetes/kubernetes/pull/62903/files#diff-f6ffeca7a3b942208d644eaa4ec99642R127 as you and @msau42 suggested. How about merging https://github.com/kubernetes/kubernetes/pull/62903 first, then use `Nsenter.HostPath` to get the path name on the host after evaluating symlinks on the host (and adding `mustExist` parameter)?